### PR TITLE
Nesting support for git-ship

### PR DESCRIPTION
**Pull request by kevgo**
@charlierudolph @allewun @ricmatsui 

Makes git-ship able to ship nested branches correctly.

Implements parts of #529

### CODE REVIEW

File: src/git-ship [hunk 1]
```diff
@@ -53,6 +67,14 @@ function steps {
   fi
   echo "delete_local_branch $target_branch_name force"
 
+  # update branch hierarchy information
+  echo delete_ancestors_entry "$target_branch_name"
+  echo delete_parent_entry "$target_branch_name"
+  child_branches "$target_branch_name" | while read branch_name; do
+    echo delete_ancestors_entry "$branch_name"
+    echo store_parent_branch "$branch_name" "$MAIN_BRANCH_NAME"
+  done
```
Comment (charlierudolph) [inline src/git-ship]: This looks very similar to what is in `git-kill` can those be dried up?

File: src/git-ship [hunk 2]
```diff
@@ -2,6 +2,19 @@
 source "$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )/helpers/helpers.sh"
 
 
+function ensure_has_no_unshipped_ancestors {
+  local branch_name=$1
+
+  local parent_branch_name ; parent_branch_name=$(parent_branch "$branch_name")
+  if [ "$parent_branch_name" != "$MAIN_BRANCH_NAME" ]; then
+    echo_error_header
+    echo_error "Shipping this branch would ship '$parent_branch_name' as well."
```
Comment (charlierudolph) [inline src/git-ship]: Should this be updated to list all ancestor branches?

File: src/git-ship (hunk 1 above)
Comment (kevgo) [inline src/git-ship]: Good idea. I have dried this up now.

File: src/git-ship (hunk 2 above)
Comment (kevgo) [inline src/git-ship]: Not a bad idea. This is a larger change, though, that requires a good amount of extra infrastructure for list formatting. Let's do this separately in #562 to keep this PR focussed. The current behavior is good enough for the being, and not wrong per se.

**Comment by kevgo** [comment]
all comments addressed

**Comment by kevgo** [comment]
It no longer deletes the remote branch if it is an ancestor to other branches. The reason is that GitHub automatically closes PRs if the branch merged into is deleted.

**Comment by kevgo** [comment]
Added a bug fix that properly removes the shipped branch from all ancestry caches. This is good to go in my opinion, and I'll ship it in 2 days unless I hear some final feedback.

**Pull request closed without merging**